ledger/src/blockstore_processor.rs

```diff
@@ -624,90 +670,127 @@ fn process_entries(
-                    if batches.is_empty() {
-                        // An entry has account lock conflicts with *itself*, which should not happen
-                        // if generated by a properly functioning leader
-                        datapoint_error!(
-                            "validator_process_entry_error",
-                            (
-                                "error",
-                                format!(
-                                    "Lock accounts error, entry conflicts with itself, txs: \
-                                     {transactions:?}"
-                                ),
-                                String
-                            )
-                        );
-                        // bail
-                        first_lock_err?;
-                    } else {
-                        // else we have an entry that conflicts with a prior entry
-                        // execute the current queue and try to process this entry again
+                queue_batches_with_lock_retry(
+                    bank,
+                    starting_index,
+                    transactions,
+                    &mut batches,
+                    |batches| {
                         process_batches(
                             bank,
                             replay_tx_thread_pool,
-                            &batches,
+                            batches,
                             transaction_status_sender,
                             replay_vote_sender,
                             batch_timing,
                             log_messages_bytes_limit,
                             prioritization_fee_cache,
-                        )?;
-                        batches.clear();
-                    }
-                }
+                        )
+                    },
+                )?;
             }
         }
     }
     process_batches(
         bank,
         replay_tx_thread_pool,
-        &batches,
+        batches.drain(..),
         transaction_status_sender,
         replay_vote_sender,
         batch_timing,
         log_messages_bytes_limit,
         prioritization_fee_cache,
     )?;
     for hash in tick_hashes {
-        bank.register_tick(hash);
+        bank.register_tick(&hash);
     }
     Ok(())
 }
 
+/// If an entry can be locked without failure, the transactions are pushed
+/// as a batch to `batches`. If the lock fails, the transactions are unlocked
+/// and the batches are processed.
+/// The locking process is retried, and if it fails again the block is marked
+/// as dead.
+/// If the lock retry succeeds, then the batch is pushed into `batches`.
+fn queue_batches_with_lock_retry(
+    bank: &Bank,
+    starting_index: usize,
+    transactions: Vec<SanitizedTransaction>,
+    batches: &mut Vec<LockedTransactionsWithIndexes<SanitizedTransaction>>,
+    mut process_batches: impl FnMut(
+        std::vec::Drain<LockedTransactionsWithIndexes<SanitizedTransaction>>,
```

Comment:
nit: `use std::vec::Drain` at the top of this file?

Reply:
061bdbdc17db4f997e7c74774f4794384f46af74